### PR TITLE
Fix: OpenEMR logs sensitive information such as payment details

### LINKED ISSUE #7340: Bug: OpenEMR logs sensitive information such as payment details
### Describe the bug
ASVS: 7.1.1 (Verify that the application does not log credentials or payment details. Session tokens should only be stored in logs in an irreversible, hashed form).
CWE: 532 (Insertion of Sensitive Information into Log File).

The OpenEMR logs check number(sensitive payment detail) when a new "Batch Payment Entry" is made, under the Category ="other" and Event ="other-insert".

### To Reproduce
1.  Login to OpenEMR using the admin credentials
2. Navigate to Fees → Batch payments to make a “Batch Payment Entry”
3. Put the “Date” as the previous date, “Post to date” as current date, “Payment Method” as Check Payment, “Check number” as “123456789”, “Payment Amount” as 100.
4. Use the default default values for “Paying entity” and “Payment category”
5. Put the “Payment From” as a user in the system, and click on the “Save changes” button.
6. Click “OK” on the confirmation and allocation pop-ups.
7. Navigate to Admin → System → Logs and click on Submit to view the logs.
8. Check for the log under Category = “other” and Event = “other-insert”.
log-sensitive-payment-detail-before-fix

### Expected behavior
The “check number” and other payment details should not be displayed(or masked) in the log

### Client configuration
- Browser: Version 122.0.6261.129 (Official Build) (arm64)
- OpenEMR version: v7.0.2
- Operating system: MacOS

### PR DESCRIPTION
**Pull request by jubittajohn**
Fixes #7340 

#### Short description of what this resolves:
This PR adds a log scrubbing template to mask sensitive details in the log(which is currently exposed in logs). Currently, this template only masks the field "reference" in the table "ar_session" to conceal sensitive payment details in the log. However, if approved, this template can be used to extend the masking of other sensitive details to other tables and fields too. Masking of sensitive information in logs is essential to prevent the stealing of information. (ASVS 7.1.1).

After update: 
log-sensitive-payment-detail-after-fix

#### Changes proposed in this pull request:
1. In "EventAuditLogger.php", add a class constant to keep track of the tables and the fields in each table that contains sensitive information.
2. In "EventAuditLogger.php", for any insert operation, if the current table contains any field containing sensitive information, the value for that field is masked to match its length.

### CODE REVIEW

File: src/Common/Logging/EventAuditLogger.php [hunk 1]
```diff
@@ -632,6 +640,27 @@ public function auditSQLEvent($statement, $outcome, $binds = null)
...
+            if ($querytype == "insert") {
+                if (strpos($comments, $table) !== false) {
+                    foreach ($fields as $field) {
+                        $pattern = "/($field)\s*=\s*'([^']+)'/";
+                        $comments = preg_replace_callback($pattern, function ($matches) {
+                            $field_name = $matches[1];
+                            $field_value = $matches[2];
+                            $masked_value = str_repeat('X', strlen($field_value));
+                            return "$field_name = '$masked_value'";
+                        }, $comments);
```
Comment (stephenwaite) [inline src/Common/Logging/EventAuditLogger.php:658]: bird_jive

Comment (jubittajohn) [reply src/Common/Logging/EventAuditLogger.php:658]: @stephenwaite is there something else that needs to be updated to the PR?

**Review by stephenwaite — APPROVED** [review]

**Comment by jubittajohn** [comment]
@stephenwaite Thank you for the approval. :)

File: src/Common/Logging/EventAuditLogger.php [hunk 2]
```diff
@@ -632,6 +640,27 @@ public function auditSQLEvent($statement, $outcome, $binds = null)
             }
         }
 
+        /**
+         * Avoid logging sensitive information in logs.
+         * For any insert operation, if the current table contains any field containing sensitive information,
+         * the value for that field is masked to match its length.
+         */
+        foreach (self::SENSITIVE_RECORDS as $table => $fields) {
```
Comment (bradymiller) [inline src/Common/Logging/EventAuditLogger.php:648]: would put the querytype == insert prior to the foreach (no reason to loop through the sensitive records array if the querytype does not equal insert)

Comment (adunsulag) [reply src/Common/Logging/EventAuditLogger.php]: @bradymiller I believe this is now resolved? Good for us to bring this in?

**Review by bradymiller — APPROVED** [review]

**Comment by bradymiller** [comment]
hi @jubittajohn , Code looks good and bringing this in. thanks for the improvement!

**Comment by jubittajohn** [comment]
@bradymiller @stephenwaite @adunsulag Thank you for looking into the Pull Request, the review comments, and the approvals.

**Comment by bradymiller** [comment]
hi @jubittajohn , turns out that modifying entries to the log is not something we should be doing to ensure OpenEMR remains compliant with ONC certification, so we will need to revert this code. Your code was well done and sorry we need to do this.

**Pull request merged**